src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -291,6 +291,12 @@ private function getAutowiredReference(TypedReference $reference)
             return new TypedReference($this->types[$type], $type);
         }
 
+        if (isset($this->types[$type])) {
+            @trigger_error(sprintf('Autowiring services based on the types they implement is deprecated since Symfony 3.3 and won\'t be supported in version 4.0. You should %s the "%s" service to "%s" instead.', isset($this->types[$this->types[$type]]) ? 'alias' : 'rename (or alias)', $this->types[$type], $type), E_USER_DEPRECATED);
```

Comment:
If somebody typehints an implementation instead of an interface, this deprecation is misleading. They do not need to change the services but also only change the typhint.